### ISSUE TITLE
Java Android API: No OpKernel was registered to support Op 'ListDiff'

### ISSUE THREAD

**Issue by andreas-eberle** [tensorflow/tensorflow#16502]
### System information
Android with Java API releases 1.4.0 and 1.5.0-rc1 (1.5 is not available yet)

### What I did
I have a graph that is applying a `tf.layers.dense` on a three dimensional tensor, which is applying a dense operation to the last dimension. Learning and execution on a Windows 10 and Ubuntu work fine.

Now I froze the model and put it on Android and receive the following error.

```
01-27 20:06:59.628 10481-11380/de.test.local W/System.err: java.util.concurrent.ExecutionException: java.lang.IllegalArgumentException: java.lang.IllegalArgumentException: No OpKernel was registered to support Op 'ListDiff' with these attrs.  Registered devices: [CPU], Registered kernels:
01-27 20:06:59.628 10481-11380/de.test.local W/System.err:   <no registered kernels>
01-27 20:06:59.628 10481-11380/de.test.local W/System.err: 	 [[Node: model/logits/Tensordot/ListDiff = ListDiff[T=DT_INT32, out_idx=DT_INT32](model/logits/Tensordot/range, model/logits/Tensordot/add_1)]]
01-27 20:06:59.628 10481-11380/de.test.local W/System.err:     at java.util.concurrent.FutureTask.report(FutureTask.java:94)
01-27 20:06:59.628 10481-11380/de.test.local W/System.err:     at java.util.concurrent.FutureTask.get(FutureTask.java:164)
01-27 20:06:59.628 10481-11380/de.test.local W/System.err:     at de.test.service.TaskWorkerLoop$Loop.run(TaskWorkerLoop.java:71)
01-27 20:06:59.628 10481-11380/de.test.local W/System.err:     at java.lang.Thread.run(Thread.java:762)
01-27 20:06:59.629 10481-11380/de.test.local W/System.err: Caused by: java.lang.IllegalArgumentException: java.lang.IllegalArgumentException: No OpKernel was registered to support Op 'ListDiff' with these attrs.  Registered devices: [CPU], Registered kernels:
01-27 20:06:59.629 10481-11380/de.test.local W/System.err:   <no registered kernels>
01-27 20:06:59.630 10481-11380/de.test.local W/System.err: 	 [[Node: model/logits/Tensordot/ListDiff = ListDiff[T=DT_INT32, out_idx=DT_INT32](model/logits/Tensordot/range, model/logits/Tensordot/add_1)]]
01-27 20:06:59.630 10481-11380/de.test.local W/System.err:     at java.lang.reflect.Constructor.newInstance0(Native Method)
01-27 20:06:59.630 10481-11380/de.test.local W/System.err:     at java.lang.reflect.Constructor.newInstance(Constructor.java:430)
01-27 20:06:59.630 10481-11380/de.test.local W/System.err:     at java8.util.concurrent.ForkJoinTask.getThrowableException(ForkJoinTask.java:565)
01-27 20:06:59.630 10481-11380/de.test.local W/System.err:     at java8.util.concurrent.ForkJoinTask.reportException(ForkJoinTask.java:646)
01-27 20:06:59.630 10481-11380/de.test.local W/System.err:     at java8.util.concurrent.ForkJoinTask.invoke(ForkJoinTask.java:704)
01-27 20:06:59.630 10481-11380/de.test.local W/System.err:     at java8.util.stream.ForEachOps$ForEachOp.evaluateParallel(ForEachOps.java:195)
01-27 20:06:59.630 10481-11380/de.test.local W/System.err:     at java8.util.stream.ForEachOps$ForEachOp$OfRef.evaluateParallel(ForEachOps.java:210)
01-27 20:06:59.630 10481-11380/de.test.local W/System.err:     at java8.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:233)
01-27 20:06:59.630 10481-11380/de.test.local W/System.err:     at java8.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:459)
01-27 20:06:59.630 10481-11380/de.test.local W/System.err:     at java8.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:624)
...
```

Is there a reason for this kernel to be missing on Android? If not, can you please add it? And what could I do in the mean time to replace the dense layer?

Many thanks in advance!

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @andrewharp: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @andrewharp: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @andrewharp: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @andrewharp: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by andreas-eberle** [comment]
any progress on this?

**Comment by tensorflowbutler** [comment]
Nagging Assignee @andrewharp: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by saudet** [comment]
@andreas-eberle The JavaCPP Presets for TensorFlow now come with the official Java API as well, and the patch you provided for Android appears to be including ListDiff, so you could try these binaries out:
http://bytedeco.org/builds/ 
https://github.com/bytedeco/javacpp-presets/tree/master/tensorflow

**Comment by tensorflowbutler** [comment]
Nagging Assignee @andrewharp: It has been 17 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @andrewharp: It has been 32 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @andrewharp: It has been 47 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @andrewharp: It has been 62 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @andrewharp: It has been 92 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @andrewharp: It has been 107 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by MusaMahmood** [comment]
I am also having this issue. Dense not working in 3-D on Android??

**Comment by bryanlimy** [comment]
I am getting the same error when using `tf.layers.dense` on 3-dimensional tensor. I tried using `tf.tensordot(tensor, weights, axes=[-1, 0])` also give me the same error on Android.

**Comment by bryanlimy** [comment]
for now, it is possible to achieve the same result by doing `layer = tf.einsum('abc,cd->abd', tensor, weights) + bias` as a single dense layer on higher dimensional tensors

**Comment by tensorflowbutler** [comment]
Nagging Assignee @andrewharp: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @andrewharp: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @andrewharp: It has been 44 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @andrewharp: It has been 59 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by MusaMahmood** [comment]
@bryanlimy @andreas-eberle This is now working with TF 1.12, so this issue may be closed, unless it still doesn't work for you.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @andrewharp: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.
